test/e2e_node/image_pull_test.go

```diff
@@ -0,0 +1,303 @@
+		var err error
+		_, is, err = getCRIClient()
+		framework.ExpectNoError(err)
+		is.SetErrorInjectors(
+			func(funcName string) error {
+				switch funcName {
+				case "PullImage":
+					time.Sleep(15 * time.Second)
+					return nil
+				default:
+					return nil
+				}
+			},
+		)
+	})
+
+	ginkgo.Context("parallel image pull with MaxParallelImagePulls=5", func() {
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.SerializeImagePulls = false
+			initialConfig.MaxParallelImagePulls = ptr.To[int32](5)
+		})
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			testpods = prepareAndCleanup(ctx, f)
+		})
+
+		ginkgo.AfterEach(func(ctx context.Context) {
+			ginkgo.By("cleanup pods")
+			for _, pod := range testpods {
+				deletePodSyncByName(ctx, f, pod.Name)
+			}
+		})
+
+		ginkgo.It("should pull immediately if no more than 5 pods", func(ctx context.Context) {
+			var pods []*v1.Pod
+			for _, testpod := range testpods {
+				pods = append(pods, e2epod.NewPodClient(f).Create(ctx, testpod))
+			}
+			for _, pod := range pods {
+				err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Running", 10*time.Minute, func(pod *v1.Pod) (bool, error) {
+					if pod.Status.Phase == v1.PodRunning {
+						return true, nil
+					}
+					return false, nil
+				})
+				framework.ExpectNoError(err)
+			}
+
+			events, err := f.ClientSet.CoreV1().Events(f.Namespace.Name).List(ctx, metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			imagePulled := map[string]*pulledStruct{}
+			// start from pulling event creationTimestamp
+			// end from pulled event creationTimestamp
+			podStartTime, podEndTime := map[string]metav1.Time{}, map[string]metav1.Time{}
+			for _, event := range events.Items {
+				if event.Reason == kubeletevents.PulledImage {
+					podEndTime[event.InvolvedObject.Name] = event.CreationTimestamp
+					for _, testpod := range testpods {
+						if event.InvolvedObject.Name == testpod.Name {
+							pulled, err := getDurationsFromPulledEventMsg(event.Message)
+							imagePulled[testpod.Name] = pulled
+							framework.ExpectNoError(err)
+							break
+						}
+					}
+				} else if event.Reason == kubeletevents.PullingImage {
+					podStartTime[event.InvolvedObject.Name] = event.CreationTimestamp
+				}
+			}
+			gomega.Expect(len(testpods)).To(gomega.BeComparableTo(len(imagePulled)))
+
+			// skip if pod1 pulling time and pod2 pulling time are not overlapped
+			if podStartTime[testpods[0].Name].Time.Before(podStartTime[testpods[1].Name].Time) {
+				if podEndTime[testpods[0].Name].Time.Before(podStartTime[testpods[1].Name].Time) {
+					e2eskipper.Skipf("pod1 pulling time and pod2 pulling time are not overlapped")
+				}
+			} else {
+				if podEndTime[testpods[1].Name].Time.Before(podStartTime[testpods[0].Name].Time) {
+					e2eskipper.Skipf("pod1 pulling time and pod2 pulling time are not overlapped")
+				}
+			}
```

Comment:
from the first look this code block(lines 100-132) is duplicated in the "Pull image" test below(lines 187-220). Can you wrap it into a func/set of funcs maybe?